Mathlib/Util/Commutes.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2023 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner
+-/
+import Lean
+import Std.Data.Option.Basic
+import Std.Tactic.Lint
+
+/-!
+
+# Auxiliary type class to avoid nested TC problems
+
+This file defines a `[@Commutes C inst₁ inst₂]` type class meaning that the
+instances `inst₁ : C` and `inst₂ : C` are defeq.  This allows us to write instances like:
```

Comment:
Presumably this is enforced socially and not by lean itself; it looks to me like the typeclass only enforces that they're propositionally equal?

Reply:
The only instance for `Commutes` witnesses definitional equality.  It's not meant to be extended.

Reply:
Oh wow that's a neat trick! I totally missed that `Commutes.mk` was an instance.